README.md
[![build status](https://secure.travis-ci.org/truepattern/aonx.png)](http://travis-ci.org/truepattern/aonx)
# aonx
Highly opinionated server framework based on node.js, express, mongoose

# Following are the standard response codes
  * http://www.iana.org/assignments/http-status-codes/http-status-codes.xml
  * 2xx - success
    * 201 - created
    * 202 - accepted
    * 204 - no content
  * 4xx - request failure
    * 400 - bad request
    * 401 - unauthorized (authentication)
    * 403 - forbidden (authorization)
    * 404 - not found
    * 405 - method not allowed
    * 406 - not acceptable
    * 409 - conflict 
    * 416 - invalid range
  * 5xx - internal server failure
    * 500 - internal server error / op timed out
    * 503 - server busy

# Features
  * Wrapper over ExpressJS, primarily targetted for an API server
  * support for http/https
  * support jsonp
    * methods PUT/DELETE/POST can be forwarded via a _method param
  * Easily connect to mongodb and have routes configured 
  * Everything is driven by simple configuration
  * Run the server in cluster mode

**ALPHA product, use at your risk, please refer to examples and todo-server**

## Todo
  * connect 2.x has compress, once available remove gzippo
  * Add mocha unit tests
  * Add authentication/authorization
  * Rate Limiting
  * API version support 
  * Usage reports
  * Monitoring
  * Clustering the server
  * Input validation ? (json/xml)


History.md
# release 0.5.2
 * upgrade to nodejs 0.8.x, express 3, mongoose 2.x

# release 0.4.8
 * added support for clustering
 * added trace on console for errors
 * removed cluster2 as express version have issues
 * moved to latest express with variable connection.encrypted set to secure
 * added authentication support for api path
 * added ignores for authentication list

# release 0.3.3
 * added support for authentication (mongoose-auth) plugins based off of mongoose
 * added https support, two additional params config.server.scheme = 'https' and config.server.key and config.server.cert (filenames) 

# release 0.2.1
 * added ObjectId support to crudHelpers
 * separated mongoose.js to mongoose-plugins

# release 0.1.6
 * added query support to index call
 * removed X-Powered-By
 * added '/' index route to return 200
 * jsonp params are filtered out
 * error codes & create method added

# release 0.1.1
 * added compression to static dir (/public)
# release 0.1.0
 * added mongoose helpers (crudHelpers)
# release 0.0.7
 * persistent session support added
 * additional configuration updates including static serving path
# release 0.0.3
 * additional examples added
# release 0.0.1
 * initial repository creation


lib/config.js
/**
 * aonx - default configuration
 */
var config = {};

// --- application configuration
config.app = {};
config.app.title = 'aonx service';
config.app.version = '0.0.1';
config.app.key = 'aonx';
config.app.secret = 'XoB_kCoL';

// cookie age would work only if db is
// enabled and set
config.app.cookieAge = 657450000;  // one week
  
// --- api related configs
config.api = {};
config.api.path = '/api/*';
config.api.keycheck = false;

// --- server related settings
config.server =  {};
config.server.port = process.env.PORT || 8074;
config.server.proxy = true;
config.server.modules = [];

// jsonp:
// - if enabled would accept a 'callback' and optional '_method'
// - _method could be PUT/DELETE/POST, so on GET with this
//   param would set the operation internally to the passed method
config.server.jsonp = false;

// path from where static files are served
config.server.path = __dirname + '/../..';
// if views is set, then 'jade' templates are enabled
//config.server.views= config.server.path + '/views';


// --- auth related settings
config.authentication = {};
config.authentication.enabled = false;
// if authentication is true, then ignores can be set as well
//config.authentication.ignores=['whoami'];
  
// --- db related settings
config.db = {};
config.db.enabled = false;
config.db.url = 'mongodb://localhost:27017/aonx';

// --- cluster configuration
config.cluster = {};
config.cluster.enabled = false;
// workers would be applicable if cluster is enabled
// by default, workers are set to # of cpus
//config.cluster.workers = 2;

// ---- export the config
module.exports = config;


lib/cluster.js
/**
 * aonx - copyright(c) 2012 truepattern.
 * MIT Licensed
 */

/** Todo's 
 * 
 */

// dependencies
var winston  = require('winston');
var _        = require('underscore');
var Cluster  = require('cluster2');

// set up module variables

// Todo
// - handle server close, close connection
// - add monPort (monitoring port)
// - add ecv


// config for cluster
//
// - server.port - where cluster would listen to
//
// 
//

var _mycluster;

/**
 * Initialize the cluster configurations
 */
exports.init = function(config) {
  var c = new Cluster({
    port: config.server.port,
    cluster: config.cluster.enabled || false,
    noWorkers: config.cluster.workers || undefined,
    ecv: {
      path: '/ecv', // Send GET to this for a heartbeat
      control: true, // send POST to /ecv/disable to disable the heartbeat, and to /ecv/enable to enable again
      monitor: '/',
      validator: function() {
        return true;
      }
    }
  });
  var processName = '['+(config.app.title || 'aonx')+'] ';
  c.on('died', function(pid) {
    winston.error(processName+'Worker ' + pid + ' died');
  });
  c.on('forked', function(pid) {
    winston.info(processName+'Worker ' + pid + ' forked');
  });
  c.on('SIGTERM', function(event) {
    winston.error(processName+'Got SIGTERM ' + JSON.stringify(event) + ' ... shutting down!');
  });
  _mycluster=c;
};


/**
 * Run the cluster
 */
exports.run = function(expressApp) {
  var c = _mycluster || new Cluster();
  c.listen(function(cb) {
    cb(expressApp);
  });
};


lib/mongoose.js
/**
 * aonx - copyright(c) 2011 truepattern.
 */

//
// CRUD helpers for mongoosejs.
//

/** Todo's 
- handle all error conditions
*/

// dependencies
var logger   = require('winston');
var _        = require('underscore');
var mongoose = require('mongoose');

// ObjectId reference
var DocumentObjectId = mongoose.Types.ObjectId;

// set up module variables
var pkgname  = '[aonx-mongoose] ';

// exported methods, variables
exports.crudHelpers = function(schema, options) {

  // json output to carry id instead of _id
  schema.method.toJSON = function (obj) {
    var o=obj.toObject(); 
    o.id=o._id; 
    delete o._id; 
    return o;
  };

  schema.statics.index = function(req,res) {
    var modelName = this.modelName.toLowerCase();
    logger.debug(pkgname + modelName + ' index called with req:'+JSON.stringify(req.query));
    var query={};
    _.extend(query,req.query);
    // lets update the query based on params passed
    // remove 'callback', '_method' & '_' (from jsonp)
    query=_.pick(query, _.without(_.keys(query),'callback', '_', '_method'));
    if(_.isEqual(query,req.query)==false) {
      logger.debug(pkgname + modelName + ' modified request:'+JSON.stringify(query));
    }
    this.find(query, function(err, objs) {
        if(!err) {
          res.send(_.map(objs, function(o) { return schema.method.toJSON(o); }));
        }
      });
  };

  schema.statics.createObject = function(req, res) {
    var model = mongoose.model(this.modelName);
    var record = new model;
    var modelName = this.modelName.toLowerCase();
    logger.debug(pkgname + modelName + ' CREATE called');
    this.createOrUpdate(req, record, function(err, newObj) {
      if(!err) {
        res.json(schema.method.toJSON(newObj), 201);
      } else {
        res.send(406);
      }
    });
  };

  schema.statics.updateObject = function(record, req, res) {
    this.createOrUpdate(req, record, function(err, newObj) {
      if(!err) {
        res.json(schema.method.toJSON(newObj), 202);
      } else {
        res.send(406);
      }
    });
  };

  schema.statics.showObject = function(record, req, res) {
    var modelName = this.modelName.toLowerCase();
    logger.debug(pkgname + modelName + ' GET called with req:'+JSON.stringify(req.query));
    if(typeof record == 'undefined') {
      logger.error(pkgname + 'invalid record passed :'+JSON.stringify(params));
    } else {
      res.json(schema.method.toJSON(record));
    }
  };

  schema.statics.removeObject = function(record, req, res) {
    var modelName = this.modelName.toLowerCase();
    logger.debug(pkgname + modelName + ' DELETE called with req:'+JSON.stringify(req.query));
    if(typeof record == 'undefined') {
      logger.error(pkgname + 'invalid record passed :'+JSON.stringify(params));
    } else {
      logger.info(pkgname + '** removing:'+JSON.stringify(record));
      record.remove(function(err) { 
          if(err) {
            // @todo: need to send appropriate error code
            logger.error(pkgname+'Error occured while removing :'+err);
            res.send(500);
          } else {
            res.send(200);
          }
        });
    }
  };

  schema.statics.findObject = function(id, cb) {
    var self = this;
    process.nextTick(function(){
      logger.debug(pkgname + 'find :'+id);
      self.findById(id, function(err, obj) {
        if(err || obj==null) {
          logger.info(pkgname + '  -- NOT found');
        } else {
          logger.debug(pkgname + '  -- found :'+JSON.stringify(obj));
        }
        cb(err, obj);
      });
    });
  };
  
  schema.statics.createOrUpdate = function(req, record, cb) {
    var body = req.body;
    /*
     * There are cases, where the whole object might
     * be passed within the modelName.
     */
    var modelName = this.modelName.toLowerCase();
    if(body && body[modelName]) {
      body = body[modelName];
    }
    var query = req.query; 
    if(query && query[modelName]) {
      query = query[modelName];
    }
    // params passed in body overrides any query params
    var params = _.extend(query, body);
    // lets update the query based on params passed
    // remove 'callback', '_method' & '_' (from jsonp)
    params=_.pick(params, _.without(_.keys(params),'callback', '_', '_method'));

    logger.debug(pkgname + modelName + ' update with :'+JSON.stringify(params));
    
    if(typeof record == 'undefined') {
      logger.error(pkgname + 'invalid record passed :'+JSON.stringify(params));
      cb('invalid record', record);
      return;
    }

    for(var p in params) {
      //@todo: raise error if the val is invalid
      if(p && params[p]) {
        var val = params[p];
        // check if this parameter is an objectId
        if(schema.paths[p] && schema.paths[p].instance=='ObjectId') {
          val = DocumentObjectId.fromString(val);
        }
        logger.debug(pkgname + '  - updating "'+p+
                     '" from "'+ record[p] + '" => "' + val +'"');
        record[p] = val;
      }
    }
    logger.info(pkgname + '** updating:'+JSON.stringify(record));
    record.save(function(err) { 
      if(err) {
        // send error
        logger.error(pkgname+'Error occured while saving :'+err);
      }
      cb(err, record);
    });
  }
};



lib/aonx.js
/**
 * aonx - copyright(c) 2011 truepattern.
 * MIT Licensed
 */

/** Todo's 
 * 
 * handle error condition on db init
 * handle assets
 * add exit handler to disconnect any connections and free resources
 * can run both http and https server at the same time?
 * set express.js flag to handle https
 */

// dependencies
var winston  = require('winston');
var express  = require('express');
var resource = require('express-resource');
var _        = require('underscore');
var Session  = require('connect-mongodb');
var gzippo   = require('gzippo');
var fs       = require('fs');
var config   = require('./config');
var http     = require('http');

// set up module variables
var pkgname  = '[aonx] ';
//winston.default.transports.console.level = 'info';
var logger = new (winston.Logger)({
  transports: [
    new (winston.transports.Console)({
      colorize: true,
      timestamp: true
    })
  ]
  //levels: config.levels,
  //colors: config.colors
});

// exported methods, variables
global.app;
global.server;

// configure the express server
exports.init = function(appconfig, preHandler) {
  logger.info(pkgname + 'initializing...');
  // if config is passed, over-ride the defaults with it
  config=_.extend(config, appconfig);
  if(config.server.scheme==='https' && config.server.key && config.server.cert) {
    var privateKey = fs.readFileSync(config.server.path+'/'+config.server.key).toString();
    var certificate = fs.readFileSync(config.server.path+'/'+config.server.cert).toString();
    global.app = express.createServer({key:privateKey, cert:certificate});
  } else {
    global.app = express();
    global.server = http.createServer(global.app);
  }
  global.config=config;
  logger.info(pkgname + '  applying config: ' + JSON.stringify(config, null,2));
  app.configure(function() {
    // call favicon first
    app.use(express.favicon());
    app.use(express.logger('dev'));
    
    // remove the powered by header - lets not give out clues
    app.use(function (req, res, next) {
      res.removeHeader("X-Powered-By");
      next();
    }); 

    // there is a bug in express.js
    // that if https is set, all the redirects
    // would not go to https, fixing
    // the variable req.connection.encrypted to force 'https'
    app.settings['trust proxy']=config.server.proxy || true;
    if(config.server.scheme==='https') {
      app.use(function(req,res,next) {
        req.connection.encrypted=true;
        next();
      });
      /*
      app.use(function(req, res, next) {
        var schema = req.headers["x-forwarded-proto"];
        // --- Do nothing if schema is already https
        if (schema === "https")
          return next();
        // --- Redirect to https
        res.redirect("https://" + req.headers.host + req.url);
      });
      */
    }

    // add the server side view engine
    if(config.server.views) {
      app.set('views', config.server.views);
      app.set('view engine', 'jade');
      // add _ as helper within template
      app.locals( { _ : require('underscore') });
      logger.info(pkgname + '  set jade as view engine...path::' + config.server.views);
    }
      
    if(config.server.jsonp) {
      app.set('jsonp callback', config.server.jsonp);
      logger.info(pkgname + '  jsonp:'+config.server.jsonp);
    }
      
    app.use(express.bodyParser());
    app.use(express.methodOverride());  
      
    app.use(express.cookieParser());
  });
  
  app.configure('development', function() {
    app.use(express.errorHandler({ dumpExceptions: true, showStack: true }));
  });
  
  app.configure('production', function() {
    app.use(express.errorHandler()); 
  });

  // lets setup the pre/post handlers
  app.configure(function() {
      // call aonx prehandler first
      _preHandlers();
      
      // call application pre-handler
      if(preHandler) {
        preHandler();
      }

      app.use(app.router);
      
      // any static files go to 'public'
      if(config.server.path) {
        //app.use(express.static(config.server.path + '/public'));
        app.use(gzippo.staticGzip(config.server.path + '/public'));
        logger.info(pkgname + '  static files served (compressed) from:'+config.server.path+'/public');
      }
      
      // register any post-handlers
      _postHandlers();
      
      // register the error handlers next to router
      _errorHandlers();
    });

};

// start the server
exports.run = function(port) {
  if(!port) {
    port=config.server.port || 8080;
  }
  logger.info(pkgname + 'starting "'+ config.app.title + 
               '" server, "' + app.settings.env + 
               '" mode on port:' + port);
  app.listen(Number(port));
  /*
  if(config.cluster.enabled) {
    cluster.init(config);
    cluster.run(app);
  } else {
    app.listen(Number(port));
  }
  */
};

// -----------------------------------------------
// internal functions


// register all pre handlers here
function _preHandlers() {
  var session;
  // enable db connection
  if(config.db.enabled) {
    var db = require('./db');
    db.init(config.db);
    logger.info(pkgname + "  mongodb initialized at:"+config.db.url);

    // export crudHelper
    exports.crudHelpers = require('mongoose-plugins').crudHelpers;
    exports.installUserSchema = require('mongoose-plugins').installUserSchema;

    // enable db sessions
    var oneWeek = 657450000;
    session = express.session({
      key    : config.app.key,
      store  : new Session({url: config.db.url }), 
      cookie : { maxAge: config.app.cookieAge||oneWeek }, 
      secret : config.app.secret
    });
    logger.info(pkgname + "  persistent sessions enabled");
  } else {
    session = express.session({
        key    : config.app.key,
        secret : config.app.secret
      });
    logger.info(pkgname + "  in-memory sessions enabled");
  }
  app.use(session);

}

// all post handlers go here
function _postHandlers() {
  // any authentication checks first
  if(config.authentication.enabled) {
    logger.info(pkgname + "  authentication to API's enabled");
    app.all(config.api.path, function(req, res, next) {
      if(config.authentication.ignores && req.params) {
        // check if the params is in ignores list
        var matches = _.filter(req.params, function(param) {
          return _.find(config.authentication.ignores, function(ignore) {
            return param.indexOf(ignore)==0;
          });
        });
        if(matches && matches.length>0) {
          logger.debug(pkgname+"ignored auth for param:" + matches);
          return next();
        }
      }
      if(req.loggedIn) {
        return next();  // auth success go to next
      } else {
        return next(error(401, 'unauthorized access'));
      }
    });
  }

  // install api checks
  if(config.api.keycheck) {
    logger.info(pkgname + "  keycheck to API's enabled");
    // here we validate the API key 
    app.all(config.api.path, function(req, res, next){
      var key = req.query[config.api.KEYNAME] || req.body[config.api.KEYNAME];
 
      // key isnt present
      if (!key) return next(error(400, 'api key required'));
 
      // key is invalid
      if (!~config.api.keys.indexOf(key)) return next(error(401, 'invalid api key'));
 
      // all good, store req.key for route access
      req.key = key;
      next();
    });
  }

  // jsonp support for delete, put & post (partial post)
  if(config.server.jsonp) {
    logger.info(pkgname + "  jsonp support on api's enabled");
    // check if callback is there along with method
    app.all(config.api.path, function(req, res, next) {
      if(req.query && req.query['callback'] && req.query['_method']) {
        // lets rewrite the method, especially PUT/DELETE/POST
        req.method=req.query['_method'];
        logger.debug(pkgname + "jsonp request converted:"+req.query);
      }
      next();
    });
  }

  var api = {
    index: function(req,res) {
      res.json({app:config.app.title, version:config.app.version});
    }
  };
  global.apiresource = app.resource('api', api);
  logger.info(pkgname + "  api handler added");
  logger.debug(pkgname + "init complete");
}


// --- Error handlers

// create an error with .status. we
// can then use the property in our
// custom error handler (Connect respects this prop as well)
function error(status, msg) {
  var err = new Error(msg);
  err.status = status;
  return err;
}

// register these error handlers in the 
// express configuration
function _errorHandlers() {
  // middleware with an arity of 4 are considered
  // error handling middleware. When you next(err)
  // it will be passed through the defined middleware
  // in order, but ONLY those with an arity of 4, ignoring
  // regular middleware.
  app.use(function(err, req, res, next){
    logger.error(err);
    console.trace();
    if(req.accepts('html') || req.is('html')) {
      res.status(err.status || 500);
      res.render('500', { error: err });
      return;
    }
    res.send(err.status || 500, { error: err.message });
  });

  // our custom JSON 404 middleware. Since it's placed last
  // it will be the last middleware called, if all others
  // invoke next() and do not respond.
  app.use(function(req, res) {
    logger.error('Failed to locate:'+req.url);
    console.trace();
    if(req.accepts('html') || req.is('html')) {
      res.status(404);
      res.render('404', { url: req.url });
      return;
    }
    res.send(404, { error: "can't find the resource" });
  });
}
